tests/testthat/test-run_now.R

```diff
@@ -139,3 +139,155 @@ test_that("Callbacks cannot affect the caller", {
+  # Exceptions in C++ callbacks
+  # =======================================================
+
+  # In these tests, in C++, later schedules a C++ callback in which an
+  # exception is thrown or interrupt occurs.
+  #
+  # Some of these callbacks in turn call R functions.
+
+  Rcpp::cppFunction(
+    depends = "later",
+    includes = '
+      #include <later_api.h>
+      #include <stdio.h>
+      #include <sys/types.h>
+      #include <unistd.h>
+      #include <signal.h>
+
+      void oof(void* data) {
+        int* v = (int*)data;
+        int value = *v;
+        delete v;
+
+        if (value == 1) {
+          throw std::runtime_error("This is a C++ exception.");
+
+        } else if (value == 2) {
+          // Throw an arbitrary object
+          throw std::string();
+
+        } else if (value == 3) {
+          // Send an interrupt to the process.
+          kill(getpid(), SIGINT);
+          sleep(3);
+
+        } else if (value == 4) {
+          // Calls R function via Rcpp, which sends interrupt signal and then
+          // sleeps. Note: This gets converted to std::runtime_error.
+          Function("r_sleep_interrupt")();
+
+        } else if (value == 5) {
+          // Calls R function via Rcpp which calls stop().
+          Function("r_error")();
+
+        } else if (value == 6) {
+          // Calls the `r_error` function via R\'s C API instead of Rcpp.
+          // Note: We don\'t actually use this for testing, because calling
+          //       Rf_eval from an Rcpp function is inherently unsafe. If an
+          //       R error occurs during the Rf_eval, a longjmp over the whole
+          //       C++ stack occurs, and the subsequent code in the function
+          //       never gets run. This function is just here to keep record of
+          //       another way that exceptions can occur.
+          SEXP e;
+          PROTECT(e = Rf_lang1(Rf_install("r_error")));
+
+          Rf_eval(e, R_GlobalEnv);
+
+          UNPROTECT(1);
+        }
+      }
+    ',
+    code = '
+      void cpp_error(int value) {
+        int* v = new int(value);
+        later::later(oof, v, 0);
+      }
+    '
+  )
+
+  r_sleep_interrupt <- function() {
+    tools::pskill(Sys.getpid(), tools::SIGINT)
+    Sys.sleep(3)
+  }
+
+  r_error <- function() {
+    stop("oopsie")
+  }
+
+  interrupted <- FALSE
+  errored     <- FALSE
+  tryCatch(
+    { cpp_error(1); run_now() },
+    error = function(e) errored <<- TRUE
+  )
+  expect_true(errored)
+
+  interrupted <- FALSE
+  error_obj <- NULL
+  tryCatch(
+    { cpp_error(2); run_now() },
+    error = function(e) errored <<- TRUE
+  )
+  expect_true(errored)
+
+  interrupted <- FALSE
+  tryCatch(
+    { cpp_error(3); run_now() },
+    interrupt = function(e) interrupted <<- TRUE
+  )
+  expect_true(interrupted)
+
+  error_obj <- NULL
+  tryCatch(
+    { cpp_error(4); run_now() },
+    error = function(e) errored <<- TRUE
+  )
+  expect_true(errored)
```

Comment:
For these `expect_true(errored)`, please reset the value of `errored` before each test.

Reply:
Oh, looks like all these `error_obj <- NULL` should be `errored <- FALSE`.

Reply:
🤦‍♂  I'll fix it.